File: examples/order/ts/src/adapters.ts
import { createServer, type IncomingMessage, type ServerResponse } from "node:http";

import { listPatternParams, type Request, type Response, type Route } from "./transport.js";

// serveOrderRoutes Mounts every Route on node:http alone.
// One Service, one Framework: the Portability Point already Lives in School.
// Reference: https://nodejs.org/api/http.html#httpcreateserveroptions-requestlistener
export async function serveOrderRoutes(routes: Route[], port: number): Promise<void> {
  const server = createServer((req, res) => dispatchOrderRequest(routes, req, res));

  await new Promise<void>((resolve) => server.listen(port, resolve));
}

// dispatchOrderRequest Matches by hand, because node:http Has no Router.
async function dispatchOrderRequest(routes: Route[], req: IncomingMessage, res: ServerResponse) {
  const url = new URL(req.url ?? "/", "http://localhost");
  const found = matchRoutePattern(routes, req.method ?? "GET", url.pathname);

  if (!found) return writeReplyAsJSON(res, { status: 404, body: { error: "no such Route" } });

  const body = await readRequestStream(req);
  writeReplyAsJSON(res, found.route.handle({
    path: found.path,
    query: Object.fromEntries(url.searchParams),
    body,
  }));
}

// matchRoutePattern is the Router node:http Does not Ship.
function matchRoutePattern(routes: Route[], method: string, pathname: string) {
  for (const route of routes) {
    const names = listPatternParams(route.pattern);
    const shape = new RegExp(`^${route.pattern.replace(/\{[^}]+\}/g, "([^/]+)")}$`);
    const found = route.method === method ? shape.exec(pathname) : null;

    if (found) {
      const path = Object.fromEntries(names.map((name, index) => [name, found[index + 1]!]));
      return { route, path };
    }
  }

  return null;
}

// readRequestStream Collects the Body node:http Delivers in Pieces.
function readRequestStream(req: IncomingMessage): Promise<string> {
  return new Promise((resolve) => {
    let text = "";
    req.on("data", (chunk) => { text += chunk; });
    req.on("end", () => resolve(text));
  });
}

// writeReplyAsJSON is the one Place that Touches a ServerResponse.
function writeReplyAsJSON(res: ServerResponse, reply: Response) {
  res.writeHead(reply.status, { "Content-Type": "application/json" });
  res.end(reply.body === null ? "" : JSON.stringify(reply.body));
}

File: examples/order/go/adapters.go
package main

import (
	"encoding/json"
	"net/http"
)

// ServeOrderRoutes Mounts every Route on a stdlib Mux.
// The only File in this Program that Imports net/http as a Framework Choice.
// One Service, one Framework: the Portability Point already Lives in School.
func ServeOrderRoutes(routes []Route, address string) error {
	mux := http.NewServeMux()

	for _, route := range routes {
		mux.HandleFunc(route.Method+" "+route.Pattern, StdlibHandlerFor(route))
	}

	return http.ListenAndServe(address, mux)
}

// StdlibHandlerFor Wraps one Route into a stdlib HandlerFunc.
func StdlibHandlerFor(route Route) http.HandlerFunc {
	names := ListPatternParams(route.Pattern)

	return func(w http.ResponseWriter, r *http.Request) {
		path := map[string]string{}
		for _, name := range names {
			path[name] = r.PathValue(name)
		}

		WriteReplyAsJSON(w, route.Handle(ReadRequestValues(r, path)))
	}
}

// WriteReplyAsJSON is the one Place that Touches a ResponseWriter.
func WriteReplyAsJSON(w http.ResponseWriter, reply Response) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(reply.Status)

	if reply.Body != nil {
		_ = json.NewEncoder(w).Encode(reply.Body)
	}
}
